### PR TITLE
MerkleRedeem distributor: allow user to claim merkle rewards to a callback contract

### PR DESCRIPTION
**Pull request by gtaschuk**
Up for discussion: Adds callbacks to the merkle distributor (as #703 does for the staking contract) to allow LPs to send rewards to a callback contract (ie for reinvestment).

I think this is most useful in the generalized use of this distributor - rather than our initially specific use case of distributing stkAAVE before the `MultiRewards` staking contract is adoptied.  Wanted to confirm my suspicion that it would be an easy extension of this contract

### CODE REVIEW

File: pkg/distributors/contracts/MerkleRedeem.sol [hunk 1]
```diff
@@ -120,6 +114,19 @@ contract MerkleRedeem is IDistributor, Ownable {
             totalBalance = totalBalance.add(claim.balance);
             claimed[claim.week][liquidityProvider] = true;
         }
+    }
+
+    /**
+     * @notice Allows a user to claim a particular week's worth of rewards
```
Comment (TomAFrench) [inline pkg/distributors/contracts/MerkleRedeem.sol:120]:
```suggestion
     * @notice Allows a user to claim a particular set of weeks' rewards
```

File: pkg/distributors/contracts/MerkleRedeem.sol [hunk 2]
```diff
@@ -128,6 +135,25 @@ contract MerkleRedeem is IDistributor, Ownable {
         }
     }
 
+    /**
+     * @notice Allows a user to claim several weeks of rewards to a callback
```
Comment (TomAFrench) [inline pkg/distributors/contracts/MerkleRedeem.sol:139]:
```suggestion
     * @notice Allows a user to claim a particular set of weeks' rewards to a callback
```

File: pkg/distributors/test/MerkleRedeem.test.ts [hunk 3]
```diff
@@ -277,8 +277,56 @@ describe('MerkleRedeem', () => {
...
+        const proof1: BytesLike[] = merkleTree1.getHexProof(elements1[0]);
+        const proof2: BytesLike[] = merkleTree2.getHexProof(elements2[0]);
+
+        claims = [
+          { week: bn(1), balance: claimBalance1, merkleProof: proof1 },
+          { week: bn(2), balance: claimBalance2, merkleProof: proof2 },
+        ];
+      });
+
+      it('allows a user to claim the reward to a callback contract', async () => {
```
Comment (nventuro) [inline pkg/distributors/test/MerkleRedeem.test.ts:302]: Let's add a test that checks that callback has been granted the tokens as internal balance

**Review by nventuro — COMMENTED** [review]
Looks great, thanks!

On an unrelated note, we've been using `.call` and reimplementing revert reason bubbling that I took the liberty of porting [`functionCall`](https://docs.openzeppelin.com/contracts/4.x/api/utils#Address-functionCall-address-bytes-) and using it here, which already does this for us. What do you think?

File: pkg/distributors/contracts/MerkleRedeem.sol [hunk 4]
```diff
@@ -128,6 +135,25 @@ contract MerkleRedeem is IDistributor, Ownable {
...
+     */
+    function claimWeeksWithCallback(
+        address payable liquidityProvider,
+        address payable callbackContract,
```
[the remark this answers is not in the document]
Comment (gtaschuk) [reply pkg/distributors/contracts/MerkleRedeem.sol:144]: Correct, though unfortunately addresses need to be marked payable to receive `IAssets`

**Comment by gtaschuk** [comment]
> Looks great, thanks!
> 
> On an unrelated note, we've been using `.call` and reimplementing revert reason bubbling that I took the liberty of porting [`functionCall`](https://docs.openzeppelin.com/contracts/4.x/api/utils#Address-functionCall-address-bytes-) and using it here, which already does this for us. What do you think?

Looks good.  Should also use for the callbacks on `MultiReward` then too

File: pkg/distributors/test/MerkleRedeem.test.ts (hunk 3 above)
Comment (gtaschuk) [reply pkg/distributors/test/MerkleRedeem.test.ts:302]: I believe that's what this test does!

**Review by nventuro — APPROVED** [review]

**Pull request merged**